### ISSUE TITLE
quest completion bug

### ISSUE THREAD

**Issue by Spongman** [Funwayguy/BetterQuesting#338]
v3.5.239

just updated lost souls modpack. all the previously completed quests now show as completed:
![image](https://user-images.githubusercontent.com/1088194/38145804-fa8c5042-33ff-11e8-8504-12e346d40d5f.png)

 but unclaimed, but when i go into the quest to try to claim it, it shows as incomplete:
![image](https://user-images.githubusercontent.com/1088194/38145767-d90cf548-33ff-11e8-802a-a06353361f10.png)

**Comment by Funwayguy** [comment]
Version 3.5.239? The current beta release is 3.5.245. Even just looking at your GUIs I can tell that they are way out of date. There are also several critical bug fixes that have been made since then including render leaks and database I/O errors in both the base mod and Standard Expansion. There was even additional functionality added to the force complete tool to clear the unclaimed status. Unfortunately just updating to the current version may not fix the weird state your quests have been put in (likely caused by a bug that was fixed in 3.5.242).

TLDR: You're running an outdated version that has issues that have since been fixed.

**Comment by Spongman** [comment]
ok. i can verify that this bug is *not* fixed in 3.5.245. the quest state is still broken as shown above. i tried restoring the save from when it was still using BQ2.

i should probably add that this is a LAN game, all users are running the same versions of everything.

i would have thought that leaving/joining a party would have helped reset this stuff, but it looks like it does nothing.

**Comment by Funwayguy** [comment]
Did you read my comment? I said that updating wouldn't fix the state your quests are now in. It only fixes the original cause so it doesn't happen again.

**Comment by Spongman** [comment]
ok, so are you saying that your mod doesn't support upgrading?

you might want to tell modpack authors not to upgrade your mod, then. seems like a huge issue, to me.

**Comment by Funwayguy** [comment]
_deep breath_

The bug was a critical issue with how true/false flags saved and loaded, some of which broke under certain conditions resetting their states and saving like that. I can't retroactively undo those changes. When it comes to updating from a 2.x version to 3.x, there are a huge number of internal format changes I cannot entirely account for and could be made worse by trying to migrate to a version with the above mentioned bug. I shouldn't have to keep reminding people to make backups before updating especially with versions which are specifically marked as BETA. I'm work with several pack developers who have no problem with this and are fully aware at the risks involved but regularly keep backups just on GitHub and know that I can usually resolve any major issue in a reasonable timeframe should the worst happen.

So I'm sorry that a bug messed up the claim status, I'm sorry that I can't retroactive undo that, and I'm sorry that I don't have a built it backup system for people who don't do it themselves.

**Comment by Spongman** [comment]
Oh, I made a backup. That's not a problem. But really, you should put a message on the curse site stating that this mod explicitly does NOT support upgrading versions - it'll save a bunch of people a lot of time.

**Comment by Spongman** [comment]
also:

> deep breath


what is this?

**Comment by Funwayguy** [comment]
If you can send me a copy of your backed up `QuestDatabase.json` and `QuestProgression.json` from before the update I'll test it in my development environment.

**Comment by Spongman** [comment]
[betterquesting.zip](https://github.com/Funwayguy/BetterQuesting/files/1866592/betterquesting.zip)


this has only ever been run in BetterQuesting-2.5.236 (from Forever+Stranded+Lost+Souls-1.0.3.7-B3)

**Comment by Funwayguy** [comment]
Weird... So I can confirm that it isn't reading the old boolean format correctly during conversion but why this only started happening now I don't know. On a good note, the one and only automated backup I _did_ implement kicked in and saved an original pre-conversion copy. I'll have to spend today digging through JSON>NBT conversion code to figure out why this is happening and get back to you.

PS: Sorry if I sounded a bit aggressive. The last few weeks has been frustrating IRL to say the least and I do still get reports for outdated versions about bugs I've long since fixed.

**Comment by Funwayguy** [comment]
Got it to load correctly again. This bug must have been there since way back I first wrote this thing and only occurs during a very specific edge case (converting a JSON boolean to NBT when there are no previous formatting numbers)

![](https://puu.sh/zUugz/2f9c81ed07.png)

Note: Don't mind the missing items, the placeholders don't have an effect on the data and code I'm testing.

**Comment by Spongman** [comment]
great! is this fixed in the latest code, or is there some simple manual transform i can do to the json to load it?

**Comment by Funwayguy** [comment]
This should be fixed in build 250 but you'll have to load the original BQ2 file to be converted again correctly.
